### PR TITLE
ISDK-2258: Update ScreenCapturerExample for TVIVideoSource

### PR DESCRIPTION
**Pull request by ceaglest**
This PR introduces a new class ExampleWebViewSource, which is a TVIVideoSource and uses the WKWebView snapshotting APIs, instead of UIGraphics. TVIScreenCapturer is still used, but only on older devices or when forced to.

Todos

* [x] Resolve conflicts with master.
* [x] Rename source file to reflect new class name.
* [x] Pixel format returned by WKWebView / TVIVideoFrame is incorrect.

### CODE REVIEW

**Comment by ceaglest** [comment]
@piyushtank this PR is ready for review. Happy to discuss any feedback tomorrow.

File: ScreenCapturerExample/ExampleWebViewSource.swift [hunk 1]
```diff
@@ -93,38 +93,99 @@ class ExampleWebViewSource: NSObject {
...
+            for row in 0 ..< cgImage.height {
+                let rowByteAddress = mutableBaseAddress.advanced(by: row * cgImage.bytesPerRow)
+
+                for pixel in stride(from: 0, to: cgImage.width * 4, by: 4) {
+                    // Swap the red and blue channels.
+                    let red = rowByteAddress[pixel]
+                    rowByteAddress[pixel] = rowByteAddress[pixel + 2]
+                    rowByteAddress[pixel+2] = red
+                }
+            }
```
Comment (ceaglest) [inline ScreenCapturerExample/ExampleWebViewSource.swift]: Consider using vImagePermuteChannels in order to swap red/blue.

https://stackoverflow.com/questions/10654909/converting-bgra-to-argb

Comment (ceaglest) [reply ScreenCapturerExample/ExampleWebViewSource.swift]: Done, and this is much faster.

**Comment by ceaglest** [comment]
I was able to transition to using vImage for RGBA to BGRA format conversion. I also had a chance to cleanup the ViewController a little bit.

I have left a todo for un-premultiplying the images on the iOS simulator, because this is crashing locally. I don't think this should prevent us from merging the PR (it's listed in the README as a known issue).

File: ScreenCapturerExample/ExampleWebViewSource.swift [hunk 2]
```diff
@@ -93,38 +94,101 @@ class ExampleWebViewSource: NSObject {
...
+                                        rowBytes: cgImage.bytesPerRow)
+
+        switch byteOrderInfo {
+        case .byteOrder32Little:
+            // Pixel format encountered on iOS simulators.
+            // Note: We have observed that pre-multiplied images might contain non-opaque alpha on the simulator.
+            assert(alphaInfo == .premultipliedFirst || alphaInfo == .noneSkipFirst)
+            if (alphaInfo == .premultipliedFirst) {
+                // TODO: This vImage call is crashing.
+//                vImageUnpremultiplyData_ARGB8888(&imageBuffer, &imageBuffer, vImage_Flags(kvImageDoNotTile))
```
Comment (piyushtank) [inline ScreenCapturerExample/ExampleWebViewSource.swift]: formatting is off. 
Whats our strategy here?

File: ScreenCapturerExample/ViewController.swift [hunk 3]
```diff
@@ -13,12 +13,12 @@ import AVFoundation
...
     var localVideoTrack: TVILocalVideoTrack?
-    var remoteView: TVIVideoView?
+    weak var localView: TVIVideoView?
 
-    /// Deprecated capturer to use on older iOS versions.
+    /// Deprecated capturer to use on older iOS versions where WKWebView snapshotting is not supported.
     var screenCapturer: TVIScreenCapturer?
 
-    /// A custom video source to capture WKWebView.
+    /// A source which uses snapshotting APIs to capture the contents of a WKWebView.
```
Comment (piyushtank) [inline ScreenCapturerExample/ViewController.swift]: s/`///`/`//`

File: ScreenCapturerExample/ExampleWebViewSource.swift (hunk 2 above)
Comment (ceaglest) [reply ScreenCapturerExample/ExampleWebViewSource.swift]: For comparison, here is the app running on my iPhone XS.

img_1715

**Comment by ceaglest** [comment]
Alright, I've went ahead and updated the README.md with my findings.

In short, yes it is possible to un-premultiply the images without crashing. On the iOS simulator you really need to make a CFMutableData or it will crash upon write. After doing all this I found that the alpha was opaque anyways and there un-premultiplying made no difference.

File: ScreenCapturerExample/ExampleWebViewSource.swift [hunk 4]
```diff
@@ -93,38 +94,100 @@ class ExampleWebViewSource: NSObject {
...
+             * Note: While UIImages captured on a device claim to have pre-multiplied alpha, the alpha channel is always opaque (0xFF).
+             */
+            assert(alphaInfo == .premultipliedLast || alphaInfo == .noneSkipLast)
+
+            // Swap the red and blue channels.
+            var permuteMap = [UInt8(2), UInt8(1), UInt8(0), UInt8(3)]
+            vImagePermuteChannels_ARGB8888(&imageBuffer,
+                                           &imageBuffer,
+                                           &permuteMap,
+                                           vImage_Flags(kvImageDoNotTile))
```
Comment (piyushtank) [inline ScreenCapturerExample/ExampleWebViewSource.swift]: 👍

File: ScreenCapturerExample/ExampleWebViewSource.swift [hunk 5]
```diff
@@ -1,10 +1,11 @@
 //
-//  ExampleScreenCapturer.swift
+//  ExampleWebViewSource.swift
 //  ScreenCapturerExample
 //
-//  Copyright © 2016-2017 Twilio, Inc. All rights reserved.
+//  Copyright © 2016-2019 Twilio, Inc. All rights reserved.
```
Comment (piyushtank) [inline ScreenCapturerExample/ExampleWebViewSource.swift]: 👍

**Pull request merged**